tests/test_coordinate_source.py

```diff
@@ -0,0 +1,328 @@
+import os
+import pickle
+import random
+import shutil
+import tempfile
+from collections import OrderedDict
+from glob import glob
+from unittest import TestCase
+
+import importlib_resources
+import mrcfile
+import numpy as np
+from click.testing import CliRunner
+from pandas import DataFrame
+
+import tests.saved_test_data
+from aspire.commands.extract_particles import extract_particles
+from aspire.noise import WhiteNoiseEstimator
+from aspire.source.coordinates import (
+    CentersCoordinateSource,
+    EmanCoordinateSource,
+    RelionCoordinateSource,
+)
+from aspire.storage import StarFile
+
+
+class CoordinateSourceTestCase(TestCase):
+    def setUp(self):
+        # temporary directory to set up a toy dataset
+        self.tmpdir = tempfile.TemporaryDirectory()
+        self.data_folder = self.tmpdir.name
+        # load pickled list of picked particle centers
+        with importlib_resources.path(
+            tests.saved_test_data, "apple_centers.p"
+        ) as centers_path:
+            # apple_centers.p was pickled with protocol 4
+            centers = pickle.load(
```

Comment:
Why use pre-computed centers? Couldn't we just randomly generate some?

Reply:
In the previous version of the PR, `test_apple.py` was also modified so the Apple-picked centers were stored in `apple_centers.p` rather than hardcoded. The advantage of using fixed coordinates here is that we know for a given size how many of them fit in the micrograph. (tested later on)

Reply:
I can give `test_apple.py` the change again too, so it's all together in one merge.

Reply:
> The advantage of using fixed coordinates here is that we know for a given size how many of them fit in the micrograph. (tested later on)

Yes but this can also be done by deliberately placing some centers close to the edge. I would argue that it's also more transparent.

Reply:
Precomputed centers replaced with randomly generated centers guaranteed to have 400 particles fit at size 256, with 100 excluded after resizing to 300.